pkg/rbd/rbd_util.go

```diff
@@ -109,36 +114,80 @@ type rbdSnapshot struct {
 
 var (
 	supportedFeatures = sets.NewString("layering")
+
+	// large interval and timeout, it should be longer than the maximum
+	// time an operation can take (until refcounting of the connections is
+	// available)
+	connPool = util.NewConnPool(15*time.Minute, 10*time.Minute)
 )
 
 // createImage creates a new ceph image with provision and volume options.
 func createImage(ctx context.Context, pOpts *rbdVolume, volSz int64, cr *util.Credentials) error {
-	var output []byte
-
-	image := pOpts.RbdImageName
 	volSzMiB := fmt.Sprintf("%dM", volSz)
+	options := librbd.NewRbdImageOptions()
 
-	logMsg := "rbd: create %s size %s format 2 (features: %s) using mon %s, pool %s "
+	logMsg := "rbd: create %s size %s (features: %s) using mon %s, pool %s "
 	if pOpts.DataPool != "" {
 		logMsg += fmt.Sprintf("data pool %s", pOpts.DataPool)
+		err := options.SetString(librbd.RbdImageOptionDataPool, pOpts.DataPool)
+		if err != nil {
+			return err
+		}
 	}
 	klog.V(4).Infof(util.Log(ctx, logMsg),
-		image, volSzMiB, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
+		pOpts.RbdImageName, volSzMiB, pOpts.ImageFeatures, pOpts.Monitors, pOpts.Pool)
 
-	args := []string{"create", image, "--size", volSzMiB, "--pool", pOpts.Pool, "--id", cr.ID, "-m", pOpts.Monitors, "--keyfile=" + cr.KeyFile, "--image-feature", pOpts.ImageFeatures}
+	if pOpts.ImageFeatures != "" {
+		// TODO: parse imagefeatures and set RbdImageOptionFeatures
+		// err = options.SetUint64(librbd.RbdImageOptionFeatures, 0)
+		klog.Infof(util.Log(ctx, "ImageFeatures (set to \"%v\" are currently "+
+			"ignored"), pOpts.ImageFeatures)
+	}
 
-	if pOpts.DataPool != "" {
-		args = append(args, "--data-pool", pOpts.DataPool)
+	ioctx, err := pOpts.getIoctx(cr)
```

Comment:
I prefer not requesting a connection per call and returning it, instead would be easier to get it once in credentials (or otherwise). Saves adding code to each call, adding the get and put wrappers.

Also, is there a plan to retain the IO context and reuse it across calls that require the same?

Reply:
I thought about putting everything in the credentials. It  might be cleaner to replace the passing of the credentials with a connection, or IOContext. That is something I want to do at a later time though.

Reply:
if we do that changes we are removing this part of code right?